### PR TITLE
Build new contracts on bootstrap, save them in config, load with checks afterward

### LINKED ISSUE #92: Consider separating parameter values from `*ContractParameters`
Now we're using them as expected constant values on contract parsing. I suggest we remove them from `*ContractParameters` since the structs are already heavy. Plus, we might introduce a new struct below.

In bootstrap, we should be able to set new `min_data_points` and other contract parameter values rather than compiling the contract with a new value to have the new constant value.

OR

re-designate them to be treated as new parameters on bootstrap or expected parameters in all other cases. Distinguish via two methods in `*Contract` (`load`/`create`?)

### PR DESCRIPTION
**Pull request by greenhat**
Close #92 

In this PR:
- Bootstrap generates a full `OracleConfig` with updated P2S in contract parameters;
- `*Contract`, `*ContractInputs`, `*BoxWrapperInputs` got two constructors: `create` (substitutes constant values in `p2s`) and `load` (checks constant values in `p2s` correspond to the properties);
- switch from storing `*ContractParameters` in `OracleConfig` to `*BoxWrapperInputs`;
- remove all addresses in `Addresses` in favor of `OracleConfig::oracle_address`;
- add `OracleConfig::rescan_height`, set during the bootstrap;

### CODE REVIEW

File: core/src/contracts/update.rs [hunk 1]
```diff
@@ -41,26 +42,58 @@ pub enum UpdateContractError {
...
+        ballot_token_id: TokenId,
+    ) -> Result<Self, UpdateContractError> {
+        let network_prefix = contract_parameters.p2s.network();
+        let inputs_to_create_contract = Self {
+            contract_parameters,
+            pool_nft_token_id,
+            ballot_token_id,
+        };
+        let refresh_contract = UpdateContract::create(&inputs_to_create_contract)?;
+        let new_parameters = refresh_contract.parameters(network_prefix);
```
Comment (kettlebell) [inline core/src/contracts/update.rs:69]:
```suggestion
        let update_contract = UpdateContract::create(&inputs_to_create_contract)?;
        let new_parameters = update_contract.parameters(network_prefix);
```

File: core/src/contracts/pool.rs [hunk 2]
```diff
@@ -35,37 +35,77 @@ pub enum PoolContractError {
...
+    ) -> Result<Self, PoolContractError> {
+        let contract_inputs = PoolContractInputs {
             contract_parameters,
-            refresh_nft_token_id: &token_ids.refresh_nft_token_id,
-            update_nft_token_id: &token_ids.update_nft_token_id,
-        }
+            refresh_nft_token_id,
+            update_nft_token_id,
+        };
+        let _refresh_contract = PoolContract::load(&contract_inputs)?;
```
Comment (kettlebell) [inline core/src/contracts/pool.rs:79]:
```suggestion
        let _ = PoolContract::load(&contract_inputs)?;
```

File: core/src/contracts/ballot.rs [hunk 3]
```diff
@@ -35,25 +36,72 @@ pub enum BallotContractError {
...
+
+    pub fn load(
+        contract_parameters: BallotContractParameters,
+        update_nft_token_id: TokenId,
+    ) -> Result<Self, BallotContractError> {
+        let contract_inputs = Self {
+            contract_parameters,
+            update_nft_token_id,
+        };
+        let _refresh_contract = BallotContract::load(&contract_inputs)?;
```
Comment (kettlebell) [inline core/src/contracts/ballot.rs:74]:
```suggestion
        let _ = BallotContract::load(&contract_inputs)?;
```

File: core/src/contracts/update.rs [hunk 4]
```diff
@@ -41,26 +42,58 @@ pub enum UpdateContractError {
...
+        contract_parameters: UpdateContractParameters,
+        pool_nft_token_id: TokenId,
+        ballot_token_id: TokenId,
+    ) -> Result<Self, UpdateContractError> {
+        let contract_inputs = Self {
+            contract_parameters,
+            pool_nft_token_id,
+            ballot_token_id,
+        };
+        let _refresh_contract = UpdateContract::load(&contract_inputs)?;
```
Comment (kettlebell) [inline core/src/contracts/update.rs:86]:
```suggestion
        let _ = UpdateContract::load(&contract_inputs)?;
```

**Review by kettlebell — CHANGES_REQUESTED** [review]
Looks great! Please check my comments for very minor issues.

Also, I wonder whether we could change the name of the `load` methods across the crate to something like `checked_load`, or `check_and_load`; just as a quick indicator to whoever reads it. What do you think?

**Comment by greenhat** [comment]
> Also, I wonder whether we could change the name of the `load` methods across the crate to something like `checked_load`, or `check_and_load`; just as a quick indicator to whoever reads it. What do you think?

I'm glad you mentioned it. I don't like `load/create` that much myself. `checked_load` sounds way better! I wonder if changing `create` to `make` or `build` or even `substitute` would better convey the idea that it's not a generic constructor and it substitutes constants in `p2s`. What do you think?

**Comment by kettlebell** [comment]
> I'm glad you mentioned it. I don't like `load/create` that much myself. `checked_load` sounds way better! I wonder if changing `create` to `make` or `build` or even `substitute` would better convey the idea that it's not a generic constructor and it substitutes constants in `p2s`. What do you think?

How about `[create | make | build]_with`? So it would be read as 'create with provided args'.

**Comment by greenhat** [comment]
Sounds good! Thanks! I like build_with.

File: core/src/contracts/update.rs (hunk 1 above)
Comment (greenhat) [reply core/src/contracts/update.rs]: Thanks for spotting this! Copy-pasting :facepalm:

**Pull request merged**